pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java

```diff
@@ -485,27 +493,57 @@ public void onUpdate(final String path, final LocalBrokerData data, final Stat s
-        brokerCandidateCache.clear();
-        LoadManagerShared.applyPolicies(serviceUnit, policies, brokerCandidateCache, loadData.getBrokerData().keySet());
-        log.info("{} brokers being considered for assignment of {}", brokerCandidateCache.size(), bundle);
-
-        // Use the filter pipeline to finalize broker candidates.
-        for (BrokerFilter filter : filterPipeline) {
-            filter.filter(brokerCandidateCache, data, loadData, conf);
-        }
-        final String broker = placementStrategy.selectBroker(brokerCandidateCache, data, loadData, conf);
+    public String selectBrokerForAssignment(final ServiceUnitId serviceUnit) {
+        // Use brokerCandidateCache as a lock to reduce synchronization.
+        synchronized(brokerCandidateCache) {
+            final String bundle = serviceUnit.toString();
+            if (preallocatedBundleToBroker.containsKey(bundle)) {
+                // If the given bundle is already in preallocated, return the selected broker.
+                return preallocatedBundleToBroker.get(bundle);
+            }
+            final BundleData data = loadData.getBundleData().computeIfAbsent(bundle, key -> getBundleDataOrDefault(bundle));
+            brokerCandidateCache.clear();
+            Set<String> activeBrokers;
+            try {
+                activeBrokers = availableActiveBrokers.get();
+            } catch (Exception e) {
+                // Try-catch block inserted because ZooKeeperChildrenCache.get throws checked exception, though we
+                // should not really see this happen unless something goes very wrong.
+                log.warn("Unexpected error when trying to get active brokers", e);
 
-        // Add new bundle to preallocated.
-        loadData.getBrokerData().get(broker).getPreallocatedBundleData().put(bundle, data);
-        preallocatedBundleToBroker.put(bundle, broker);
-        return broker;
+                // Fall back to using loadData key set.
+                activeBrokers = loadData.getBrokerData().keySet();
+            }
+            LoadManagerShared.applyPolicies(serviceUnit, policies, brokerCandidateCache, activeBrokers);
+            log.info("{} brokers being considered for assignment of {}", brokerCandidateCache.size(), bundle);
+
+            // Use the filter pipeline to finalize broker candidates.
+            for (BrokerFilter filter : filterPipeline) {
+                filter.filter(brokerCandidateCache, data, loadData, conf);
+            }
+            final String broker = placementStrategy.selectBroker(brokerCandidateCache, data, loadData, conf);
+
+            // Add new bundle to preallocated.
+            loadData.getBrokerData().get(broker).getPreallocatedBundleData().put(bundle, data);
+            preallocatedBundleToBroker.put(bundle, broker);
+
+            // Make sure broker is still active.
+            try {
+                if (!availableActiveBrokers.get().contains(broker)) {
```

Comment:
I think we can remove this double check of broker that is part of `availableActiveBrokers` because we have already verified at line-508. So, we can remove it and immediately return derived broker.